MSB4181: The "Restore Task" task returned false but did not log an error.

### NuGet Product Used

dotnet.exe, MSBuild.exe, NuGet.exe, Visual Studio Package Management UI

### Product Version

Repro'd at least back to 17.9

### Worked before?

Yes, not sure

### Impact

It's more difficult to complete my work

### Repro Steps & Context

Particular restore errors are no longer being logged, so it's very difficult for customers to know why restore is failing.
In my case, I can repro this by adding a package reference for a package which exists on a feed, but I'm not authorized to download.

1. Add an Azure DevOps feed to your config.
2. Configure an upstream feed which requires authentication.
3. Remove any credentials that would authenticate your account on the upstream feed.
4. Add a PackageReference for a package version that is on the Upstream feed but not the Main feed.
5. Attempt to restore - even with detailed verbosity - and the following error results:
  `msbuild src\NuGet.Clients\NuGet.PackageManagement.PowerShellCmdlets\NuGet.PackageManagement.PowerShellCmdlets.csproj /t:restore /clp:v=n`
    > error MSB4181: The "Restore Task" task returned false but did not log an error.

The actual error I found while debugging, which I expect to be logged, is:

> NuGet.Protocol.Core.Types.FatalProtocolException: 'The feed 'vside [https://pkgs.dev.azure.com/azure-public/vside/_packaging/msft_consumption/nuget/v3/index.json]' lists package 'Microsoft.Internal.VisualStudio.Shell.Framework.17.8.37385' but multiple attempts to download the nupkg have failed. The feed is either invalid or required packages were removed while the current operation was in progress. Verify the package exists on the feed and try again.'

> Unable to find package 'Microsoft.Internal.VisualStudio.Shell.Framework.17.8.37385'.
>    at NuGet.Protocol.FindPackagesByIdNupkgDownloader.<GetNuspecReaderFromNupkgAsync>d__7.MoveNext() in C:\NuGet.Client\src\NuGet.Core\NuGet.Protocol\Utility\FindPackagesByIdNupkgDownloader.cs:line 90
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
>    at NuGet.Protocol.HttpFileSystemBasedFindPackageByIdResource.<GetDependencyInfoAsync>d__12.MoveNext() in C:\NuGet.Client\src\NuGet.Core\NuGet.Protocol\RemoteRepositories\HttpFileSystemBasedFindPackageByIdResource.cs:line 200
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
>    at NuGet.Commands.SourceRepositoryDependencyProvider.<GetDependenciesCoreAsync>d__26.MoveNext() in C:\NuGet.Client\src\NuGet.Core\NuGet.Commands\RestoreCommand\SourceRepositoryDependencyProvider.cs:line 369
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
>    at NuGet.TaskResultCache`2.<>c__5`1.<GetOrAddAsync>b__5_1(Task`1 task) in C:\NuGet.Client\build\Shared\TaskResultCache.cs:line 98
>    at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
>    at System.Threading.Tasks.Task.Execute()


### Verbose Logs

_No response_

--- a/src/NuGet.Core/NuGet.DependencyResolver.Core/ResolverUtility.cs
+++ b/src/NuGet.Core/NuGet.DependencyResolver.Core/ResolverUtility.cs
@@ -91,6 +91,7 @@ namespace NuGet.DependencyResolver
                                                     Strings.Error_PackageNotFoundWhenExpected,
                                                         match.Provider.Source,
                                                     ex.PackageIdentity.ToString());
+                        context.Logger.LogError(message);
 
                         throw new FatalProtocolException(message, ex);
                     }

--- a/test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/FindPackageTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/FindPackageTests.cs
@@ -1,8 +1,8 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
-using System;
 using System.Collections.Generic;
+using System.Globalization;
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
@@ -176,6 +176,7 @@ namespace NuGet.DependencyResolver.Core.Tests
             var token = CancellationToken.None;
             var edge = new GraphEdge<RemoteResolveResult>(null, null, null);
             var actualIdentity = new LibraryIdentity("x", NuGetVersion.Parse("1.0.0-beta"), LibraryType.Package);
+            var packageIdentity = new PackageIdentity(actualIdentity.Name, actualIdentity.Version);
             var dependencies = new[] { new LibraryDependency() { LibraryRange = new LibraryRange("y", VersionRange.All, LibraryDependencyTarget.Package) } };
             var dependencyInfo = LibraryDependencyInfo.Create(actualIdentity, framework, dependencies);
 
@@ -198,6 +199,12 @@ namespace NuGet.DependencyResolver.Core.Tests
 
             // Assert
             Assert.Equal(2, hitCount);
+            Assert.Equal(1, testLogger.Errors);
+            string errorMessage = string.Format(CultureInfo.CurrentCulture,
+                                                Strings.Error_PackageNotFoundWhenExpected,
+                                                remoteProvider.Object.Source,
+                                                packageIdentity.ToString());
+            Assert.Equal(errorMessage, testLogger.ErrorMessages.Single());
         }
 
         [Fact]

--- a/test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/FindPackageTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/FindPackageTests.cs
@@ -165,7 +165,7 @@ namespace NuGet.DependencyResolver.Core.Tests
         }
 
         [Fact]
-        public async Task FindPackage_VerifyMissingListedPackageThrowsNotFoundAndLogsErrorAsync()
+        public async Task FindPackage_VerifyMissingListedPackage_ThrowsNotFoundAndLogsErrorAsync()
         {
             // Arrange
             var range = new LibraryRange("x", VersionRange.Parse("1.0.0-beta"), LibraryDependencyTarget.Package);

--- a/test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/FindPackageTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/FindPackageTests.cs
@@ -165,7 +165,7 @@ namespace NuGet.DependencyResolver.Core.Tests
         }
 
         [Fact]
-        public async Task FindPackage_VerifyMissingListedPackageThrowsNotFound()
+        public async Task FindPackage_VerifyMissingListedPackageThrowsNotFoundAndLogsErrorAsync()
         {
             // Arrange
             var range = new LibraryRange("x", VersionRange.Parse("1.0.0-beta"), LibraryDependencyTarget.Package);

